### ISSUE TITLE
No result after running "extractUnrelated_wrapper.R"

### ISSUE THREAD

**Issue by liu-xu-bing** [rounakdey/FastSparseGRM#10]
Hi

I used plink1.9 to generate .fam file with filtering MAF > 0.05 and LD pruned variants and then used king=2.2.7 with command --ibdseg to generate .seg file. The .fam contained 660 samples but .seg file only had 13 rows. Then I ran "getDivergence_wrapper.R" and generated .div file successfully. But after I ran the "extractUnrelated_wrapper.R", the result file contained no result and there was no error message in "extractUnrelated.Rout". Could you help me to solve this problem?

Here is the info in "extractUnrelated.Rout" file:
------------------------------------------------------------------------------------------------
R version 3.6.3 (2020-02-29)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /home/liuxubing/miniconda3/envs/staar/lib/libopenblasp-r0.3.12.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] FastSparseGRM_1.01 optparse_1.6.6

loaded via a namespace (and not attached):
 [1] tidyselect_1.1.1       purrr_0.3.4            splines_3.6.3
 [4] operator.tools_1.6.3   lattice_0.20-44        vctrs_0.3.8
 [7] generics_0.1.0         GWASExactHW_1.01       getopt_1.20.3
[10] stats4_3.6.3           mgcv_1.8-35            utf8_1.2.1
[13] rlang_1.1.1            nloptr_1.2.2.2         pillar_1.6.1
[16] glue_1.4.2             DBI_1.1.1              BiocGenerics_0.32.0
[19] gdsfmt_1.22.0          GMMAT_1.4.0            GenomeInfoDbData_1.2.2
[22] foreach_1.5.1          lifecycle_1.0.3        zlibbioc_1.32.0
[25] Biostrings_2.54.0      codetools_0.2-18       Biobase_2.46.0
[28] SeqArray_1.26.0        IRanges_2.20.0         GenomeInfoDb_1.22.0
[31] parallel_3.6.3         fansi_0.4.2            broom_0.7.6
[34] Rcpp_1.0.11            backports_1.2.1        S4Vectors_0.24.0
[37] XVector_0.26.0         lme4_1.1-27            formula.tools_1.7.1
[40] dplyr_1.0.6            CompQuadForm_1.4.3     GenomicRanges_1.38.0
[43] grid_3.6.3             cli_3.6.1              bitops_1.0-7
[46] magrittr_2.0.1         RCurl_1.98-1.3         tibble_3.1.2
[49] mice_3.13.0            SNPRelate_1.20.0       crayon_1.4.1
[52] tidyr_1.1.3            pkgconfig_2.0.3        ellipsis_0.3.2
[55] MASS_7.3-54            Matrix_1.3-3           data.table_1.14.0
[58] minqa_1.2.4            assertthat_0.2.1       logistf_1.24
[61] iterators_1.0.13       boot_1.3-28            R6_2.5.0
[64] SeqVarTools_1.24.0     igraph_1.2.6           nlme_3.1-150
[67] compiler_3.6.3
> extractUnrelated(file.seg,file.div,prefix.in,degree,file.include,prefix.out)
[1] "Extracting the set of unrelated samples"
[1] "9% completed"
[1] "13% completed"
[1] "17% completed"
[1] "22% completed"
[1] "26% completed"
[1] "35% completed"
[1] "39% completed"
[1] "43% completed"
[1] "52% completed"
[1] "61% completed"
[1] "70% completed"
[1] "74% completed"
[1] "83% completed"
[1] "91% completed"
[1] "100% completed"
>
> proc.time()
   user  system elapsed
  7.650   3.020  21.655
-----------------------------------------------------------------------------------------------------------

Best
Xubing

**Comment by rounakdey** [comment]
Hi Xubing,

Could you please see if there is some `.unrels` file generated? That should be the output.
Also, since your sample-size is relatively small (660), I would advice not to use LD pruned SNPs for the KING IBDsegment step. For large sample-sizes (>10K) using LD pruned SNPs is good enough, but may not be accurate if the sample-size is relatively small.

**Comment by liu-xu-bing** [comment]
Hi Runakdey

I have used unfiltered .bed file to run the tools and generated .unrels file but still no result in this file.

**Comment by rounakdey** [comment]
Can you please let me know the file names of the `file.seg` and `file.div` argument used for `extractUnrelated_wrapper.R`? This is to check whether the inputs were correct. And please include the extensions in the file names.

**Comment by liu-xu-bing** [comment]
OK. Here are the commands I used:
1. VCF to bed
`plink --allow-extra-chr -vcf high_myopia_and_normal_sample_VQSR_hail_PASS_exclude_non_variants.vcf.gz --make-bed --out high_myopia`
2. king
`king -b high_myopia.bed --ibdseg --degree 4 --cpus 10 --prefix high_myopia`
3. getDivergence
`R CMD BATCH --vanilla '--args --prefix.in high_myopia --file.seg high_myopia.seg --num_threads 20 --prefix.out high_myopia_ibdseg_divergence' ~/software/FastSparseGRM/extdata/getDivergence_wrapper.R getDivergence.Rout`
4. extractUnrelated
`R CMD BATCH --vanilla '--args --prefix.in high_myopia --file.seg high_myopia.seg --degree 4 --file.div high_myopia_ibdseg_divergence.div --prefix.out high_myopia_ibdseg_unrelated' ~/software/FastSparseGRM/extdata/extractUnrelated_wrapper.R extractUnrelated_ibdseg.Rout`

Here are intermediate files:
1. high_myopia.seg : 
FID1    ID1     FID2    ID2     IBD1Seg IBD2Seg PropIBD InfType
D1907008301     D1907008301     D1907008547     D1907008547     0.0031  0.9927  0.9943  Dup/MZ
D1907008586     D1907008586     D1907008668     D1907008668     0.0019  0.9923  0.9933  Dup/MZ
D1907008603     D1907008603     D1907008604     D1907008604     0.0000  0.9967  0.9967  Dup/MZ
D2001000981     D2001000981     D2001000982     D2001000982     0.2655  0.0000  0.1328  3rd
DTDP1908101916-1        DTDP1908101916-1        DTDP1908101951-1        DTDP1908101951-1        0.1920  0.0000  0.0960  3rd
DTDP1908105342-1        DTDP1908105342-1        DTDP1908105376-1        DTDP1908105376-1        0.9873  0.0000  0.4937  PO
RDPYD17100021   A       RDPYD17100023   A       0.9751  0.0012  0.4888  PO
RDPYD17100022   A       RDPYD17100023   A       0.9739  0.0000  0.4870  PO
RDPYD18012955   A       RDPYD18012956   A       0.1100  0.0015  0.0565  4th
RDPYD18034260   B       RDPYD18034261   B       0.4903  0.2155  0.4607  FS
RDPYD18184157   B       RDPYD19040676   C       0.1497  0.0000  0.0748  4th
RDPYD19022756   A       RDPYD19022757   A       0.2484  0.0000  0.1242  3rd

2. high_myopia_ibdseg_divergence.div
D1907008301 D1907008301 9
D1907008547 D1907008547 7
D1907008586 D1907008586 9
D1907008603 D1907008603 8
D1907008604 D1907008604 8
D1907008668 D1907008668 9
D2001000981 D2001000981 4
D2001000982 D2001000982 8
DTDP1908101916-1 DTDP1908101916-1 2
DTDP1908101951-1 DTDP1908101951-1 3
DTDP1908105342-1 DTDP1908105342-1 3
DTDP1908105376-1 DTDP1908105376-1 3
RDPYD17100021 A 3
RDPYD17100022 A 5
RDPYD17100023 A 4
RDPYD18012955 A 3
RDPYD18012956 A 6
RDPYD18034260 B 6
RDPYD18034261 B 7
RDPYD18184157 B 7
RDPYD19022756 A 4
RDPYD19022757 A 7
RDPYD19040676 C 6

3. getDivergence.Rout
> getDivergence(file.seg,prefix.in,num_threads,divThresh,degree,nRandomSNPs,prefix.out)
[1] "Calculating ancestry divergence for the samples in the seg file"
Reading the file: /picb/lilab6/datasets/restricted/yanke/high_myopia/snp/bim/high_myopia.bed
Reading BED file complete
Number of SNPs read: 36311380
Number of samples: 660
Using 20 threads
Processing samples complete
5% completed in 21.831s
>
>
> proc.time()
   user  system elapsed
437.899 145.436 229.694

**Comment by rounakdey** [comment]
Thanks for sending the information. I can see what happened there. FastSparseGRM requires unique IIDs, it basically selects samples based on IIDs only. There are many subjects in your dataset with the same IID but different FIDs (eg. `RDPYD17100021 A`, and `RDPYD18012956 A`) which is creating the problem. I will suggest the following, please edit the *.fam, *.seg, and *.div files and just make the FID and IIDs the same and unique for each subject. For example, in the previous example RDPYD17100021 A can be edited as to having FID and IID both `RDPYD17100021-A`. This FID, IID differentiation is still there to have things kind of backwards compatible with plink. However, most modern genetic analysis software have done with this differentiation and only uses one unique subject ID. So, to match with them, we only use the IID to identify the subjects, the FID is mainly ignored. You can safely edit the *.fam, *.seg, *.div files without worrying about redoing any of the analysis. Once you have the edited versions, you can run the `extractUnrelated_wrapper.R` again, and it should interpret the files correctly then.

**Comment by liu-xu-bing** [comment]
Thank you for your recommendation. I changed the sample ID from "_" to "-" and have successfully created the "high_myopia_rename_ibdseg_degree4_unrelated.unrels". But when I ran the next step "Run PCA", I got new error: 

Allele frequencies calculated
[1] "Reading Completed in 6 seconds"
[1] "1 iterations completed"
[1] "2 iterations completed"
[1] "3 iterations completed"
[1] "4 iterations completed"
[1] "5 iterations completed"
[1] "6 iterations completed"
[1] "7 iterations completed"
[1] "8 iterations completed"
[1] "9 iterations completed"
[1] "10 iterations completed"
[1] "Iterations Completed in 147 seconds"
Error in svd(H, nv = 0) : infinite or missing values in 'x'
Calls: runPCA -> drpca -> svd
Execution halted

I have tested both unfiltered bed and filtered bed but the error still appeared. Could you help me to solve this problem? I will be eternally grateful.

**Comment by rounakdey** [comment]
Hi, this is a very common error. This error occurs if you have any variant in the bed file which is monomorphic (MAC=0) in the unrelated set of subjects. Please filter the bed file for the PCA (and the GRM calculation) step based on MAC (>5) and it should run fine.

**Comment by liu-xu-bing** [comment]
I'm sorry for too long time to reply you. I have tested the variants sets with both MAC >3 and MAC > 5. The MAC >3 variants set didn't work and the MAC >5 set didn't throw the previous error but throw the new error:

> runPCA(prefix.in,file.unrels,no_pcs,num_threads,prefix.out,no_iter)
[1] "PCA started at 2023-10-08 09:37:56"
[1] "11 relateds, 649 unrelateds, and 12682676 variants"
Reading the file: /picb/lilab6/datasets/restricted/yanke/high_myopia/snp/bim/high_myopia_exclude_rename_mac5.bed
Reading BED file complete
Number of SNPs read: 12682676
Number of samples: 660
Using 20 threads
Processing samples complete
Allele frequencies calculated
[1] "Reading Completed in 8 seconds"
[1] "1 iterations completed"
[1] "2 iterations completed"
[1] "3 iterations completed"
[1] "4 iterations completed"
[1] "5 iterations completed"
[1] "6 iterations completed"
[1] "7 iterations completed"
[1] "8 iterations completed"
[1] "9 iterations completed"
[1] "10 iterations completed"
[1] "Iterations Completed in 195 seconds"
Error: std::bad_alloc
Execution halted

**Comment by liu-xu-bing** [comment]
I'm sorry for so long time to reply you. I have tested both variants sets with MAC > 3 and MAC >5 and found that the MAC > 3 set didn't work but the MAC > 5 set run well. What's more, when running the step "Run PCA" with all the MAC >5  variants, an error was throwed as "std::bad_alloc". I think it is because my vcf was generated from WGS dataset. So I filtered MAF > 0.05 variants by plink and run all the steps successfully. Thank you again for your help.
